solutions/swb-reference/src/services/swbProjectService.ts

```diff
@@ -0,0 +1,287 @@
+  }
+
+  /**
+   * List projects
+   *
+   * @param request - the request object for listing projects
+   * @returns Project entries in DDB, with optional pagination token
+   */
+  public async listProjects(request: ListProjectsRequest): Promise<PaginatedResponse<Project>> {
+    const getUserGroupsResponse = await this._dynamicAuthorizationService.getUserGroups({
+      authenticatedUser: request.user,
+      userId: request.user.id
+    });
+    return this._projectService.listProjects(request, getUserGroupsResponse.data.groupIds);
+  }
+
+  /**
+   * Soft deletes a project from the database.
+   *
+   * @param request - a {@link DeleteProjectRequest} object that contains the id of the project to delete
+   * @param checkDependencies - an async function that checks if there are dependencies associated with the project
+   */
+  public async softDeleteProject(
+    request: DeleteProjectRequest,
+    checkDependencies: (projectId: string) => Promise<void>
+  ): Promise<void> {
+    await this._projectService.softDeleteProject(request, checkDependencies);
+
+    const authenticatedUser = request.authenticatedUser;
+    const projectId: string = request.projectId;
+    const paRole: string = getProjectAdminRole(projectId);
+    const researcherRole: string = getResearcherRole(projectId);
+
+    // Get PA Permissions to delete
+    const paIdentityPermissions: IdentityPermission[] = this._generatePAIdentityPermissions(
+      projectId,
+      paRole
+    );
+
+    // Get Researcher Permissions to delete
+    const researcherIdentityPermissions: IdentityPermission[] = this._generateResearcherIdentityPermissions(
+      projectId,
+      researcherRole
+    );
+
+    const deleteIdentityPermissionsRequest = DeleteIdentityPermissionsRequestParser.parse({
+      authenticatedUser,
+      identityPermissions: [...paIdentityPermissions, ...researcherIdentityPermissions]
+    });
+
+    await this._dynamicAuthorizationService.deleteIdentityPermissions(deleteIdentityPermissionsRequest);
+
+    await this._dynamicAuthorizationService.deleteGroup({
+      authenticatedUser,
+      groupId: paRole
+    });
+    await this._dynamicAuthorizationService.deleteGroup({
+      authenticatedUser,
+      groupId: researcherRole
+    });
+  }
+
+  /**
+   * Update the name or description of an existing project.
+   *
+   * @param request - a {@link UpdateProjectRequest} object that contains the id of the project to update
+   *                  as well as the new field values
+   * @returns a {@link Project} object that reflects the changes requested
+   */
+  public async updateProject(request: UpdateProjectRequest): Promise<Project> {
+    return this._projectService.updateProject(request);
+  }
+
+  /**
+   *
+   */
+  private _generatePAIdentityPermissions(projectId: string, paRole: string): IdentityPermission[] {
+    return [
+      ...this._generateIdentityPermissions('*', SwbAuthZSubject.SWB_DATASET, ['CREATE', 'READ'], paRole, {
```

Comment:
idk whether laying all this out in a literal is crazy or genius lol

Reply:
lol I thought this would provide the best readability, definitely open to other suggestions.